ssl/s3_lib.c

```diff
@@ -3574,16 +3579,79 @@ long ssl3_ctrl(SSL *s, int cmd, long larg, void *parg)
         break;
 
     case SSL_CTRL_GET_TLSEXT_STATUS_REQ_OCSP_RESP:
-        *(unsigned char **)parg = sc->ext.ocsp.resp;
-        if (sc->ext.ocsp.resp_len == 0
-                || sc->ext.ocsp.resp_len > LONG_MAX)
-            return -1;
-        return (long)sc->ext.ocsp.resp_len;
+        *(unsigned char **)parg = NULL;
+        ret = -1;
+
+#ifndef OPENSSL_NO_OCSP
+         resp = sk_OCSP_RESPONSE_value(sc->ext.ocsp.resp_ex, 0);
+
+        if (resp != NULL) {
+            int resp_len = i2d_OCSP_RESPONSE(resp, NULL);
+
+            if (resp_len > 0 && resp_len <= LONG_MAX) {
+                OPENSSL_free(sc->ext.ocsp.resp);
+
+                sc->ext.ocsp.resp = OPENSSL_malloc(resp_len);
+
+                if (sc->ext.ocsp.resp != NULL) {
+                    *(unsigned char **)parg = sc->ext.ocsp.resp;
+                    p = sc->ext.ocsp.resp;
+                    sc->ext.ocsp.resp_len = i2d_OCSP_RESPONSE(resp, &p);
+                    ret = sc->ext.ocsp.resp_len;
+                } else {
+                    ret = -1;
+                }
+            }
```

Comment:
If you initialize `p` to `NULL`, this is much more simple and correct:
```suggestion
            int resp_len = i2d_OCSP_RESPONSE(resp, &p);

            if (resp_len > 0) {
                OPENSSL_free(sc->ext.ocsp.resp);
                *(unsigned char **)parg = sc->ext.ocsp.resp = p;
                ret = sc->ext.ocsp.resp_len = resp_len;
            } else {
                ret = -1;
            }
```